research/conditional/utils/model_utils.py

```diff
@@ -207,12 +207,31 @@ def get_expert_choice_args(args):
         and args.topk_fraction is not None
         and args.n_experts is not None
     ) and (args.effective_dff is None and args.total_experts_width is None)
+    set_arguments_option3 = (  # this should be the default
+        args.granularity is not None
+        and args.expansion_rate is not None
+        and args.dff_x is not None
+        and args.effective_dff_x is not None
+    )
 
-    if not set_arguments_option1 and not set_arguments_option2:
+    if (
+        not set_arguments_option1
+        and not set_arguments_option2
+        and not set_arguments_option3
+    ):
         raise AssertionError(
-            "You must specify either total_experts_width, effective_dff, and n_experts or expert_size, topk_fraction, and n_experts"
+            "You must specify either total_experts_width, effective_dff, and n_experts "
+            "or expert_size, topk_fraction, and n_experts "
+            "or granularity, expansion_rate, and effective_dff_x "
         )
 
+    if set_arguments_option3:
+        # 4 is the standard dff_x, we assume it's defined relative to that
+        dff_x = 4
```

Comment:
This variable is a mystery to me. If it comes from the paper, maybe we can add a comment what it does above. If not, could we rename it?